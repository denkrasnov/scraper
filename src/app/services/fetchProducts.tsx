import React, {
  useEffect,
  useReducer,
  useContext,
  createContext,
  Dispatch,
  FC
} from "react";
import axios from "axios";

import {
  Action,
  State,
  ActionTypes,
  ProductsProviderProps,
  Product
} from "./types";

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case ActionTypes.FETCH:
      return {
        ...state,
        query: action.payload,
        isLoading: true,
        isError: ""
      };
    case ActionTypes.FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: ""
      };
    case ActionTypes.FETCH_ERROR:
      return {
        ...state,
        isLoading: false,
        isError: action.payload
      };
    case ActionTypes.FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        products: action.payload
      };
    default:
      throw new Error();
  }
};

const initialState = {
  isLoading: false,
  isError: "",
  query: "",
  products: null
};

const fetchProducts = (): [State, Dispatch<Action>] => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const { query } = state;

  useEffect(() => {
    let didCancel = false;

    const fetchData = async () => {
      dispatch({ type: ActionTypes.FETCH_INIT });
      try {
        const result: {
          data: { products: Product[] | null };
        } = await axios.post("/search", { query }); // TODO": GET and type data

        if (!didCancel) {
          dispatch({
            type: ActionTypes.FETCH_SUCCESS,
            payload: result.data.products
          });
        }
      } catch (error) {
        if (!didCancel) {
          dispatch({ type: ActionTypes.FETCH_ERROR, payload: error });
        }
        // eslint-disable-next-line no-console
        console.error("Something went wrong:", error);
      }
    };

    if (query) {
      fetchData();
    }

    return () => {
      didCancel = true;
    };
  }, [query]);

  return [state, dispatch];
};

interface MainReducer {
  products: [State, Dispatch<Action>];
}

export const StateContext = createContext<MainReducer>({
  products: [initialState, () => {}]
});

const mainReducer = () => ({
  products: fetchProducts()
});

export const ProductsProvider: FC<ProductsProviderProps> = ({ children }) => (
  <StateContext.Provider value={mainReducer()}>
    {children}
  </StateContext.Provider>
);

export const useFullContext = () => useContext(StateContext);

export default ProductsProvider;
