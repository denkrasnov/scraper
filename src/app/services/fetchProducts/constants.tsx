export const INITIAL_STATE = {
  isLoading: false,
  isError: "",
  query: "",
  products: null
};
