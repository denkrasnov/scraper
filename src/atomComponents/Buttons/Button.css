.Button {
  min-width: 100px;
  padding: 12px;
  border-radius: 4px;
  background-color: #5A6169;
  color: #fff;
  font-weight: 700;
}
