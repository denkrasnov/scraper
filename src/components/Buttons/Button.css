.Button {
  min-width: 100px;
  padding: 12px;
  border-radius: 4px;
  background-color: var(--button-secondary);
  color: var(--white);
  font-weight: 700;
}
