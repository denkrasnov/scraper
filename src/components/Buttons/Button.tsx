import React, { FC } from "react";
import { Props } from "./types";

const Button: FC<Props> = props => {
  const { children, onClick } = props;
  return (
    <button type="button" onClick={onClick}>
      {children}
    </button>
  );
};

export default Button;
