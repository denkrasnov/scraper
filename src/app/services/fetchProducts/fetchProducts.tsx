import { useEffect, useReducer, Dispatch } from "react";
import axios from "axios";

import { INITIAL_STATE } from "./constants";
import reducer from "./reducer";
import { Action, State, ActionTypes, Product } from "./types";

const fetchProducts = (): [State, Dispatch<Action>] => {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  const { query } = state;

  useEffect(() => {
    let didCancel = false;

    const fetchData = async () => {
      dispatch({ type: ActionTypes.FETCH_INIT });

      try {
        const result: {
          data: { products: Product[] | null };
        } = await axios.post("/search", { query }); // TODO": GET and type data

        if (!didCancel) {
          dispatch({
            type: ActionTypes.FETCH_SUCCESS,
            payload: result.data.products
          });
        }
      } catch (error) {
        if (!didCancel) {
          dispatch({ type: ActionTypes.FETCH_ERROR, payload: error });
        }
        // eslint-disable-next-line no-console
        console.log("Something went wrong:", error);
      }
    };

    if (query) {
      fetchData();
    }

    return () => {
      didCancel = true;
    };
  }, [query]);

  return [state, dispatch];
};

export default fetchProducts;
