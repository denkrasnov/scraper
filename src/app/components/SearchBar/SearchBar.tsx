import React, { FC } from "react";

import Box from "../../../components/Box";
import Button from "../../../components/Buttons";
import Input from "../../../components/Input";

const SearchBar: FC = () => {
  return (
    <>
      <Input placeholder="Search" />
      <Box padding="0 10px">
        <Button>search</Button>
      </Box>
    </>
  );
};

export default SearchBar;
