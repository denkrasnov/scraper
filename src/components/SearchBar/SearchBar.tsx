import React, { FC } from "react";
import { Form, Field } from "react-final-form";

import Box from "src/atoms/Box";
import Button from "src/atoms/Buttons";
import Input from "src/atoms/Input";
import { useFullContext } from "src/services/ContextProvider";
import { ActionTypes } from "src/services/fetchProducts/types";

const SearchBar: FC = () => {
  const [, dispatch] = useFullContext();

  return (
    <Form
      onSubmit={values => {
        dispatch({ type: ActionTypes.FETCH, payload: values.search.trim() });
      }}
    >
      {({ handleSubmit }) => {
        return (
          <form onSubmit={handleSubmit} style={{ width: "100%" }}>
            <Box margin="auto" maxWidth="739px" width="100%">
              <Field
                name="search"
                validate={value => (value && value.trim() ? undefined : true)}
              >
                {({ input }) => {
                  return (
                    <Input
                      name={input.name}
                      onChange={input.onChange}
                      placeholder="Search"
                      value={input.value}
                    />
                  );
                }}
              </Field>
              <Box padding="0 0 0 10px">
                <Button type="submit">Search</Button>
              </Box>
            </Box>
          </form>
        );
      }}
    </Form>
  );
};

export default SearchBar;
