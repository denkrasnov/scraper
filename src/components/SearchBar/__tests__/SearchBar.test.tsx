import React from "react";
import { mount } from "enzyme";

import Button from "src/atoms/Buttons";
import * as AppContext from "src/services/ContextProvider";
import { INITIAL_STATE } from "src/services/fetchProducts/constants";
import SearchBar from "..";

describe("SearchBar", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it("should render", () => {
    jest
      .spyOn(AppContext, "useFullContext")
      .mockImplementation(() => [INITIAL_STATE, () => {}]);

    const component = mount(<SearchBar />);

    expect(component).toMatchSnapshot();
  });

  it("should submit the Form", () => {
    const dispatchMock = jest.fn();
    jest
      .spyOn(AppContext, "useFullContext")
      .mockImplementation(() => [INITIAL_STATE, dispatchMock]);

    const component = mount(<SearchBar />);

    const button = component.find(Button);
    button.simulate("click");

    expect(dispatchMock).toHaveBeenCalledTimes(1);
  });
});
