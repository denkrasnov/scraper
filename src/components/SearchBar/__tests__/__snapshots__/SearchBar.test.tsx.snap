// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render 1`] = `
<SearchBar>
  <Box
    maxWidth="739px"
    width="100%"
  >
    <div
      className="Box"
      style={
        Object {
          "maxWidth": "739px",
          "width": "100%",
        }
      }
    >
      <ReactFinalForm
        onSubmit={[Function]}
      >
        <Field
          name="search"
        >
          <Input
            name="search"
            onChange={[Function]}
            placeholder="Search"
            value=""
          >
            <input
              className="Input"
              name="search"
              onChange={[Function]}
              placeholder="Search"
              type="text"
              value=""
            />
          </Input>
        </Field>
        <Box
          padding="0 0 0 10px"
        >
          <div
            className="Box"
            style={
              Object {
                "padding": "0 0 0 10px",
              }
            }
          >
            <Button
              onClick={[Function]}
            >
              <button
                className="Button"
                onClick={[Function]}
                type="button"
              >
                search
              </button>
            </Button>
          </div>
        </Box>
      </ReactFinalForm>
    </div>
  </Box>
</SearchBar>
`;
